File: seidel/bounding_box.py
from copy import copy

from reprit.base import generate_repr

from .point import Point


class BoundingBox:
    __slots__ = 'empty', 'lower', 'upper'

    def __init__(self, empty: bool, lower: Point, upper: Point) -> None:
        self.empty = empty
        self.lower = lower
        self.upper = upper

    __repr__ = generate_repr(__init__)

    def __eq__(self, other: 'BoundingBox') -> bool:
        return ((self.empty is other.empty
                 and self.lower == other.lower
                 and self.upper == other.upper)
                if isinstance(other, BoundingBox)
                else NotImplemented)

    def add(self, point: Point) -> None:
        if self.empty:
            self.empty = False
            self.lower = copy(point)
            self.upper = copy(point)
        else:
            if point.x < self.lower.x:
                self.lower.x = point.x
            elif point.x > self.upper.x:
                self.upper.x = point.x
            if point.y < self.lower.y:
                self.lower.y = point.y
            elif point.y > self.upper.y:
                self.upper.y = point.y

    def expand(self, delta: Point) -> None:
        if not self.empty:
            self.lower -= delta
            self.upper += delta

File: src/bounding_box.cpp
#include "bounding_box.h"

BoundingBox::BoundingBox(bool empty_, const Point& lower_, const Point& upper_)
    : empty(empty_), lower(lower_), upper(upper_) {}

void BoundingBox::add(const Point& point) {
  if (empty) {
    empty = false;
    lower = upper = point;
  } else {
    if (point.x < lower.x)
      lower.x = point.x;
    else if (point.x > upper.x)
      upper.x = point.x;

    if (point.y < lower.y)
      lower.y = point.y;
    else if (point.y > upper.y)
      upper.y = point.y;
  }
}

void BoundingBox::expand(const Point& delta) {
  if (!empty) {
    lower -= delta;
    upper += delta;
  }
}
